### PR TITLE
IGNITE-15806 CPP: Remove JNI module and move it to Core.

### PR DESCRIPTION
**Pull request by ivandasch**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/platforms/cpp/core/include/ignite/jni/utils.h [hunk 1]
```diff
@@ -124,8 +119,16 @@ namespace ignite
...
+            void Init(const common::concurrent::SharedPointer<java::JniContext>& ctx0, jobject obj0) {
+                ctx = ctx0;
+
+                if (ctx.IsValid())
+                    this->obj = ctx.Get()->Acquire(obj0);
+            }
+
             /** Context. */
-            java::JniContext* ctx;
+            common::concurrent::SharedPointer<java::JniContext> ctx;
```
Comment (isapego) [inline modules/platforms/cpp/core/include/ignite/jni/utils.h:131]: Are you sure there is no cyclic reference?

**Review by isapego — COMMENTED** [review]
Overall, looks good. Please, find my comments inline.

File: modules/platforms/cpp/core/include/ignite/jni/utils.h (hunk 1 above)
Comment (ivandasch) [reply modules/platforms/cpp/core/include/ignite/jni/utils.h:131]: I suppose there is no.

1. JniContext doesn't have any references to JavaGlobalRef
2. The only declared instance of JavaGlobalRef belongs to IgniteEnvironment, the same for shared_ptr to JniContext.

Therefore, there is no cyclic reference and all resources will be freed during destruction of IgniteEnvironment.

May be I am wrong, but I suppose that all ok.

Comment (ivandasch) [reply modules/platforms/cpp/core/include/ignite/jni/utils.h:131]: I've created small test to simulate our situation

```
#include <memory>
#include <iostream>

using namespace std;


class Ctx {
public:
    ~Ctx() {
        cout << "destroyed";
    }
};

class Wrapper{
public:
    Wrapper() {}
    Wrapper(shared_ptr<Ctx> pCtx_): pCtx(pCtx_) {}
private:
    shared_ptr<Ctx> pCtx;
};

class Env {
public:
    Env(shared_ptr<Ctx> pCtx_): pCtx(pCtx_) {}
    void AddWrapper() {
        cout << "Num references before setting wrapper " << pCtx.use_count() << endl;
        this->wrapper = Wrapper(pCtx);
        cout << "Num references after after setting wrapper " << pCtx.use_count() << endl;
    }
private:
    shared_ptr<Ctx> pCtx;
    Wrapper wrapper;
};

int main() {
    auto pCtx = make_shared<Ctx>();
    cout << "Num references before assignment " << pCtx.use_count() << endl;
    {
        Env env(pCtx);
        env.AddWrapper();
        env.AddWrapper();
    }
    cout << "Num references final " << pCtx.use_count() << endl;
    return 0;
}
```
Num references before assignment 1
Num references before setting wrapper 2
Num references after after setting wrapper 3
Num references before setting wrapper 3
Num references after after setting wrapper 3
Num references final 1
destroyed
```

File: modules/platforms/cpp/core/CMakeLists.txt [hunk 2]
```diff
@@ -17,12 +17,15 @@
...
 
+find_library(JVM_LIBRARY jvm ${JAVA_JVM_LIBRARY_DIRECTORIES})
+
 include_directories(SYSTEM ${JNI_INCLUDE_DIRS})
 include_directories(include)
 
 set(TARGET ${PROJECT_NAME})
 
 set(SOURCES src/ignite.cpp
+		src/jni/java.cpp
```
[the remark this answers is not in the document]
Comment (ivandasch) [reply modules/platforms/cpp/core/CMakeLists.txt]: Here and in other files changed `\t` to whitespaces

**Review by isapego — APPROVED** [review]

**Pull request closed without merging**